packages/@react-aria/selection/src/useSelectableItem.ts

```diff
@@ -165,12 +175,14 @@ export function useSelectableItem(options: SelectableItemOptions): SelectableIte
     // Otherwise, have selection happen onPress (prevents listview row selection when clicking on interactable elements in the row)
     if (!allowsDifferentPressOrigin) {
       itemPressProps.onPress = (e) => {
-        if (e.pointerType !== 'keyboard') {
-          onSelect(e);
-        }
+        if (hasPrimaryAction || (hasSecondaryAction && e.pointerType !== 'mouse')) {
+          if (e.pointerType === 'keyboard' && !isActionKey()) {
```

Comment:
how can we have a keyboard onPress that isn't an action key?

Reply:
Action key = Enter, Selection Key = Space

Reply:
Ah, yep, ok, weird thing then
https://reactspectrum.blob.core.windows.net/reactspectrum/b0e0b1d9f96ae8a0823015f9b74510194bb65205/storybook/index.html?path=/story/listview--selectionstyle-checkbox-onaction
If i select using space, then deselect using space. It triggers action when it deselects

Reply:
+1 to what Rob saw. Something similar also happens if the user selects a row via clicking the checkbox and clicks on the row to deselect it

Reply:
It should now only trigger onSelectionChange when de-selecting, not onAction. And when an onAction is defined, the Space key always selects, and the Enter key always only performs actions, not selection. Actions cannot be performed once in selection mode, matching mouse/touch.